torch/distributed/_sharded_tensor/api.py

```diff
@@ -750,3 +796,61 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
                           device=local_device, )
     else:
         raise ValueError(f'Unsupported create_op: {tensor_init_params.create_op}')
+
+
+class PartialTensor(object):
+    """
+    PartialTensor is an abstraction to represent Tensors that need
+    aggregation across multiple devices and multiple processes.
+
+    PartialTensor is initialized in an SPMD like fashion where each rank
+    initializes the PartialTensor. The PartialTensor object on each rank
+    then only stores the local partial shard and provides global metadata
+    for all the shards.
+
+    PartialTensor doesn't provide any Tensor like operations but is a
+    wrapper providing the Tensor representing the local partial shard
+    and the global metadata.
+
+    Using these, users can apply custom distributed sharded computations
+    on top of this primitive. The local shards are all initialized using
+    the create_op specified by tensor_init_params.create_op, e.g.,
+    torch.ones, or torch.empty.
+
+    Args:
+        local_partial_results (List(Tensor)):
```

Comment:
What are the semantics here? If we aggregate do we get a List[Tensor] of aggregated results by aggregating each index? I think it is simpler to just allow only a single Tensor per rank.

Reply:
So the local result is a list of tensors. After aggregation, each rank will have one tensor because we will use `reduce_scatter`. Using one tensor also works, we then need to stack or concat them into a single tensor. We can then specify the dim in the resharding spec. Let me change it.